src/lib.ts

```diff
@@ -0,0 +1,320 @@
+			(args.quality as number) < 0 ||
+			(args.quality as number) > 100
+		) {
+			delete args.quality;
+			errors.push('quality arg is not valid');
+		}
+	}
+	if (args.resize) {
+		if (!/^\d+(px)?,\d+(px)?$/.test(args.resize as string)) {
+			delete args.resize;
+			errors.push('resize arg is not valid');
+		}
+	}
+	if (args.crop_strategy) {
+		if (!/^(smart|entropy|attention)$/.test(args.crop_strategy)) {
+			delete args.crop_strategy;
+			errors.push('crop_strategy arg is not valid');
+		}
+	}
+	if (args.gravity) {
+		if (!/^(north|northeast|east|southeast|south|southwest|west|northwest|center)$/.test(args.gravity)) {
+			delete args.gravity;
+			errors.push('gravity arg is not valid');
+		}
+	}
+	if (args.fit) {
+		if (!/^\d+(px)?,\d+(px)?$/.test(args.fit as string)) {
+			delete args.fit;
+			errors.push('fit arg is not valid');
+		}
+	}
+	if (args.crop) {
+		if (!/^\d+(px)?,\d+(px)?,\d+(px)?,\d+(px)?$/.test(args.crop as string)) {
+			delete args.crop;
+			errors.push('crop arg is not valid');
+		}
+	}
+	if (args.zoom) {
+		if (!/^\d+(\.\d+)?$/.test(args.zoom)) {
+			delete args.zoom;
+			errors.push('zoom arg is not valid');
+		}
+	}
+	if (args.webp) {
+		if (!/^0|1|true|false$/.test(args.webp as string)) {
+			delete args.webp;
+			errors.push('webp arg is not valid');
+		}
+	}
+	if (args.lb) {
+		if (!/^\d+(px)?,\d+(px)?$/.test(args.lb)) {
+			delete args.lb;
+			errors.push('lb arg is not valid');
+		}
+	}
+	if (args.background) {
+		if (!/^#[a-f0-9]{3}[a-f0-9]{3}?$/.test(args.background)) {
+			delete args.background;
+			errors.push('background arg is not valid');
+		}
+	}
+
+	// crop (assumes crop data from original)
+	if (args.crop) {
+		let cropValuesString = typeof args.crop === 'string' ? args.crop.split(',') : args.crop;
+
+		// convert percentages to px values
+		let cropValues = cropValuesString.map(function (value, index) {
+			if (value.indexOf('px') > -1) {
+				return Number(value.substring(0, value.length - 2));
+			} else {
+				return Number(
+					Number((metadata[index % 2 ? 'height' : 'width'] as number) * (Number(value) / 100)).toFixed(0)
+				);
+			}
+		});
+
+		image.extract({
+			left: cropValues[0],
+			top: cropValues[1],
+			width: cropValues[2],
+			height: cropValues[3],
+		});
+	}
+
+	// get zoom value
+	const zoom = parseFloat(args.zoom || '1') || 1;
+
+	// resize
+	if (args.resize) {
+		// apply smart crop if available
+		if (args.crop_strategy === 'smart' && !args.crop) {
+			const cropResize = getDimArray(args.resize);
+			const rotatedImage = await image.toBuffer();
+			const result = await smartcrop.crop(rotatedImage, {
+				width: cropResize[0] as number,
+				height: cropResize[1] as number,
+			});
+
+			if (result && result.topCrop) {
+				image.extract({
+					left: result.topCrop.x,
+					top: result.topCrop.y,
+					width: result.topCrop.width,
+					height: result.topCrop.height,
+				});
+			}
+		}
+
+		// apply the resize
+		args.resize = getDimArray(args.resize, zoom) as number[];
```

Comment:
Why are you needing to cast these everywhere?

Reply:
getDimArray can also return null. In most of these places I didn't want to fully refactor as I didn't want to introduce more bugs, so it's going what it did before, but mostly I'm also trying to get the types to pass. Technically there is some subtle behavior bugs that already existed but I don't really want to go rewrite it all